src/DSROracleBase.sol

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.0;
+
+import { IDSROracle } from './interfaces/IDSROracle.sol';
+
+/**
+ * @title  DSROracleBase
+ * @notice Base functionality for all DSR oracles.
+ */
+abstract contract DSROracleBase is IDSROracle {
+
+    uint256 private constant RAY = 1e27;
+
+    IDSROracle.PotData internal _data;
+
+    function getPotData() external view returns (IDSROracle.PotData memory) {
+        return _data;
+    }
+
+    function getDSR() external view returns (uint256) {
+        return _data.dsr;
+    }
+
+    function getChi() external view returns (uint256) {
+        return _data.chi;
+    }
+
+    function getRho() external view returns (uint256) {
+        return _data.rho;
+    }
+
+    function getAPR() external view returns (uint256) {
+        unchecked {
+            return (_data.dsr - RAY) * 365 days;
+        }
+    }
+
+    function getConversionRate() external view returns (uint256) {
+        return getConversionRate(block.timestamp);
+    }
+
+    function getConversionRate(uint256 timestamp) public view returns (uint256) {
+        IDSROracle.PotData memory d = _data;
+        uint256 rho = d.rho;
+        if (timestamp == rho) return d.chi;
+        require(timestamp >= rho, "DSROracleBase/invalid-timestamp");
+        
+        return (timestamp > rho) ? _rpow(d.dsr, timestamp - rho) * uint256(d.chi) / RAY : d.chi;
+    }
+
+    function getConversionRateBinomialApprox() external view returns (uint256) {
+        return getConversionRateBinomialApprox(block.timestamp);
+    }
+
+    // Copied and slightly modified from https://github.com/aave/aave-v3-core/blob/42103522764546a4eeb856b741214fa5532be52a/contracts/protocol/libraries/math/MathUtils.sol#L50
+    function getConversionRateBinomialApprox(uint256 timestamp) public view returns (uint256) {
+        IDSROracle.PotData memory d = _data;
+        uint256 rho = d.rho;
+        if (timestamp == rho) return d.chi;
+        require(timestamp >= rho, "DSROracleBase/invalid-timestamp");
+        
+        uint256 exp;
+        uint256 rate;
+        unchecked {
+            exp = timestamp - rho;
+            rate = d.dsr - RAY;
+        }
+
+        uint256 expMinusOne;
+        uint256 expMinusTwo;
+        uint256 basePowerTwo;
+        uint256 basePowerThree;
+        unchecked {
+            expMinusOne = exp - 1;
+
+            expMinusTwo = exp > 2 ? exp - 2 : 0;
+
+            basePowerTwo = rate * rate / RAY;
+            basePowerThree = basePowerTwo * rate / RAY;
+        }
```

Comment:
Why are we dividing by seconds per year here, isn't rate a per second rate?

Reply:
Are you asking why we aren't diving by seconds per year while the original code is? The answer is yes DSR is a per second rate already and Aave rate is in yearly amount.

Reply:
Oh shit sorry I totally mixed up the diff there haha. Pasted the aave code in and it was green my bad